src/main/java/net/openhft/chronicle/core/io/VanillaReferenceCounted.java

```diff
@@ -158,10 +157,17 @@ public void throwExceptionIfNotReleased() throws IllegalStateException {
 
     @Override
     public void warnAndReleaseIfNotReleased() throws ClosedIllegalStateException {
-        if (refCount() > 0) {
-            if (!unmonitored && !AbstractCloseable.DISABLE_DISCARD_WARNING)
-                Slf4jExceptionHandler.WARN.on(type, "Discarded without being released");
-            callOnRelease();
+        for (; ; ) {
+            final int refCount = refCount();
+            if (refCount == 0) {
+                break;
+            }
+            if (valueCompareAndSet(refCount, 0)) {
+                if (!unmonitored && !AbstractCloseable.DISABLE_DISCARD_WARNING)
+                    Jvm.warn().on(type, "Discarded without being released");
+                callOnRelease();
+                break;
+            }
```

Comment:
This extra code reduces the chance of showing a warning when you want it to detect incorrect behaviour as early as possible.

Reply:
If anything, you want to more reliably show a warning than avoid it.

Reply:
Good point, I'll change that.